### PR TITLE
[INTERNAL] Tungsten Cache code review and refactoring

### PR DESCRIPTION
**Pull request by JoshRosen**
@yhuai, @feynmanliang, I decided to open this PR against my own repo so that we have a place to discuss tungsten cache code.

[Review on Reviewable](https://reviewable.io/reviews/joshrosen/spark/5)

- [ ] Add test for caching empty partition

### CODE REVIEW

File: sql/core/src/main/scala/org/apache/spark/sql/SQLImplicits.scala [hunk 1]
```diff
@@ -110,4 +121,181 @@ private[sql] abstract class SQLImplicits {
...
+     * @param compressionType "" (default), "lz4", "lzf", or "snappy", see
+     *                        [[CompressionCodec.ALL_COMPRESSION_CODECS]]
+     * @param blockSize size of each MemoryBlock (default = 4 MB)
+     */
+    def tungstenCache(
+        compressionType: String = "", blockSize: Int = 4000000): (RDD[_], DataFrame) = {
+      val schema = df.schema
+
+      val convert = CatalystTypeConverters.createToCatalystConverter(schema)
+      val internalRows = df.rdd.map(convert(_).asInstanceOf[InternalRow])
```
Comment (JoshRosen) [inline sql/core/src/main/scala/org/apache/spark/sql/SQLImplicits.scala]: This is going to be moderately inefficient if the input happens to already be in UnsafeRow format, since this will end up performing an extra round of conversions, converting each UnsafeRow into a public row, then turning around and converting it back into an UnsafeRow.

Instead, I think that it would make more sense to write a function which takes an InternalRow and returns an UnsafeRow but which acts as a no-op when applied to existing UnsafeRows.

File: sql/core/src/main/scala/org/apache/spark/sql/SQLImplicits.scala [hunk 2]
```diff
@@ -110,4 +121,181 @@ private[sql] abstract class SQLImplicits {
...
+      val cachedRDD = internalRows.mapPartitions { rowIterator =>
+        val bufferedRowIterator = rowIterator.buffered
```
Comment (JoshRosen) [inline sql/core/src/main/scala/org/apache/spark/sql/SQLImplicits.scala]: Why is this buffering necessary?

File: sql/core/src/main/scala/org/apache/spark/sql/SQLImplicits.scala [hunk 3]
```diff
@@ -110,4 +121,181 @@ private[sql] abstract class SQLImplicits {
...
+        val convertToUnsafe = UnsafeProjection.create(schema)
+        val taskMemoryManager = new TaskMemoryManager(SparkEnv.get.executorMemoryManager)
```
Comment (JoshRosen) [inline sql/core/src/main/scala/org/apache/spark/sql/SQLImplicits.scala]: I think that we should get the TaskMemoryManager from TaskContext instead of creating a new one here.

File: sql/core/src/main/scala/org/apache/spark/sql/SQLImplicits.scala [hunk 4]
```diff
@@ -110,4 +121,181 @@ private[sql] abstract class SQLImplicits {
...
+        val compressionCodec: Option[CompressionCodec] = if (compressionType.isEmpty) {
+          None
+        } else {
+          Some(CompressionCodec.createCodec(SparkEnv.get.conf, compressionType))
+        }
+
+        new Iterator[MemoryBlock] {
+          // NOTE: This assumes that size of every row < blockSize
```
Comment (JoshRosen) [inline sql/core/src/main/scala/org/apache/spark/sql/SQLImplicits.scala]: We will have to eventually figure out how to handle large rows.

File: sql/core/src/main/scala/org/apache/spark/sql/SQLImplicits.scala [hunk 5]
```diff
@@ -110,4 +121,181 @@ private[sql] abstract class SQLImplicits {
...
+     * Packs the rows of [[df]] into contiguous blocks of memory.
+     * @param compressionType "" (default), "lz4", "lzf", or "snappy", see
+     *                        [[CompressionCodec.ALL_COMPRESSION_CODECS]]
+     * @param blockSize size of each MemoryBlock (default = 4 MB)
+     */
+    def tungstenCache(
+        compressionType: String = "", blockSize: Int = 4000000): (RDD[_], DataFrame) = {
+      val schema = df.schema
+
+      val convert = CatalystTypeConverters.createToCatalystConverter(schema)
```
Comment (JoshRosen) [inline sql/core/src/main/scala/org/apache/spark/sql/SQLImplicits.scala]: `convert` is an ambiguous name; I'd prefer `convertToCatalyst`.

File: sql/core/src/main/scala/org/apache/spark/sql/SQLImplicits.scala [hunk 6]
```diff
@@ -110,4 +121,181 @@ private[sql] abstract class SQLImplicits {
...
+            // Packs rows into a `blockSize` bytes contiguous block of memory, starting a new block
+            // whenever the current fills up
+            // Each row is laid out in memory as [rowSize (4)|row (rowSize)]
+            val block = taskMemoryManager.allocateUnchecked(blockSize)
+
+            var currOffset = 0
+            val seenRows = ArrayBuffer[InternalRow]()
+            while (bufferedRowIterator.hasNext && currOffset < blockSize) {
+              val currRow = convertToUnsafe.apply(bufferedRowIterator.head)
+              val recordSize = 4 + currRow.getSizeInBytes
```
Comment (JoshRosen) [inline sql/core/src/main/scala/org/apache/spark/sql/SQLImplicits.scala]: Add comment to explain that the extra four bytes are for storing size.

File: sql/core/src/main/scala/org/apache/spark/sql/SQLImplicits.scala [hunk 7]
```diff
@@ -110,4 +121,181 @@ private[sql] abstract class SQLImplicits {
...
+              if (currOffset + recordSize < blockSize) {
+                Platform.putInt(
+                  block.getBaseObject, block.getBaseOffset + currOffset, currRow.getSizeInBytes)
+                currRow.writeToMemory(block.getBaseObject, block.getBaseOffset + currOffset + 4)
+                seenRows += currRow
```
Comment (JoshRosen) [inline sql/core/src/main/scala/org/apache/spark/sql/SQLImplicits.scala]: Don't we need to perform a defensive copy here if we're going to buffer these unsafe rows, since the row returned from the projection will be backed by a re-used buffer inside of that projection?

File: sql/core/src/main/scala/org/apache/spark/sql/SQLImplicits.scala [hunk 8]
```diff
@@ -110,4 +121,181 @@ private[sql] abstract class SQLImplicits {
...
+                bufferedRowIterator.next()
+              }
+              currOffset += recordSize // Increment currOffset regardless to break loop when full
+            }
+
+            // Optionally compress block before writing
+            compressionCodec match {
+              case Some(codec) =>
+                // Compress the block using an on-heap byte array
+                val blockArray = new Array[Byte](blockSize)
```
Comment (JoshRosen) [inline sql/core/src/main/scala/org/apache/spark/sql/SQLImplicits.scala]: Given that the block size is fixed, I think that it would make sense to re-use the on-heap compression buffer across batches.

File: sql/core/src/main/scala/org/apache/spark/sql/SQLImplicits.scala [hunk 9]
```diff
@@ -110,4 +121,181 @@ private[sql] abstract class SQLImplicits {
...
+        override val needConversion = false
+
+        override def buildScan(): RDD[Row] = {
+          val numFields = this.schema.length
+          val _compressionType: String = compressionType
+          val _blockSize = blockSize
+
+          cachedRDD.flatMap { rawBlock =>
+            // Optionally decompress block
+            val compressionCodec: Option[CompressionCodec] = if (_compressionType.isEmpty) {
```
Comment (JoshRosen) [inline sql/core/src/main/scala/org/apache/spark/sql/SQLImplicits.scala]: Might be cleaner to make compressionType into an option?

File: sql/core/src/main/scala/org/apache/spark/sql/SQLImplicits.scala [hunk 10]
```diff
@@ -110,4 +121,181 @@ private[sql] abstract class SQLImplicits {
...
+                moreData = false
+              }
+            }
+            rows
+          }.asInstanceOf[RDD[Row]]
+        }
+
+        override def toString: String = getClass.getSimpleName + s"[${df.toString}]"
+      }
+      cachedRDD.count // trigger caching action because it's lazy
```
Comment (JoshRosen) [inline sql/core/src/main/scala/org/apache/spark/sql/SQLImplicits.scala]: Wait, don't we want this new cache to be lazy as well?

File: sql/core/src/test/scala/org/apache/spark/sql/CachedTableSuite.scala [hunk 11]
```diff
@@ -341,6 +341,31 @@ class CachedTableSuite extends QueryTest with SharedSQLContext {
     }
   }
 
+  test("tungsten cache uncompressed table and read") {
```
Comment (JoshRosen) [inline sql/core/src/test/scala/org/apache/spark/sql/CachedTableSuite.scala]: Could reduce some duplication here by using a loop / ScalaTest's ability to have programatic test definitions.

File: sql/core/src/main/scala/org/apache/spark/sql/SQLImplicits.scala [hunk 12]
```diff
@@ -110,4 +121,181 @@ private[sql] abstract class SQLImplicits {
...
+              case None => rawBlock
+            }
+
+            val rows = new ArrayBuffer[InternalRow]()
+            var currOffset = 0
+            var moreData = true
+            while (currOffset < block.size() && moreData) {
+              val rowSize = Platform.getInt(block.getBaseObject, block.getBaseOffset + currOffset)
+              currOffset += 4
+              // TODO: should probably have a null terminator rather than relying on zeroed out
```
Comment (JoshRosen) [inline sql/core/src/main/scala/org/apache/spark/sql/SQLImplicits.scala]: Could store `-1` for the row size; I think we do this in other places, too.

File: sql/core/src/main/scala/org/apache/spark/sql/SQLImplicits.scala [hunk 13]
```diff
@@ -110,4 +121,181 @@ private[sql] abstract class SQLImplicits {
...
+//                    cumNumRead,
+//                    uncompressedBlockArray.length - cumNumRead)
+//                }
+//                cis.close()
+                cdis.close()
+                MemoryBlock.fromByteArray(uncompressedBlockArray)
+              case None => rawBlock
+            }
+
+            val rows = new ArrayBuffer[InternalRow]()
```
Comment (JoshRosen) [inline sql/core/src/main/scala/org/apache/spark/sql/SQLImplicits.scala]: Couldn't you return an iterator of rows which streams through the block to deserialize the rows lazily? I wonder whether buffering all of these rows is going to be expensive + involve lots of unnecessary object creation.

File: sql/core/src/main/scala/org/apache/spark/sql/SQLImplicits.scala (hunk 2 above)
Comment (feynmanliang) [inline sql/core/src/main/scala/org/apache/spark/sql/SQLImplicits.scala]: L166 calls `head` in order to see if we can fit this row into the current `MemoryBlock`; we want to start a new `MemoryBlock` without advancing the iterator if it doesn't fit

File: sql/core/src/main/scala/org/apache/spark/sql/SQLImplicits.scala [hunk 14]
```diff
@@ -110,4 +121,181 @@ private[sql] abstract class SQLImplicits {
...
   }
+
+  /**
+   * ::Experimental::
+   *
+   * Pimp my library decorator for tungsten caching of DataFrames.
+   * @since 1.5.1
+   */
+  @Experimental
+  implicit class TungstenCache(df: DataFrame) {
```
Comment (JoshRosen) [inline sql/core/src/main/scala/org/apache/spark/sql/SQLImplicits.scala]: For code cleanliness, I wonder whether we should move the bulk of the logic out of this class and into its own file.

File: core/src/main/scala/org/apache/spark/io/CompressionCodec.scala [hunk 15]
```diff
@@ -106,7 +106,7 @@ private[spark] object CompressionCodec {
 class LZ4CompressionCodec(conf: SparkConf) extends CompressionCodec {
 
   override def compressedOutputStream(s: OutputStream): OutputStream = {
-    val blockSize = conf.getSizeAsBytes("spark.io.compression.lz4.blockSize", "32k").toInt
```
Comment (JoshRosen) [inline core/src/main/scala/org/apache/spark/io/CompressionCodec.scala]: Going to roll this change back.

File: sql/core/src/main/scala/org/apache/spark/sql/SQLImplicits.scala (hunk 12 above)
Comment (feynmanliang) [inline sql/core/src/main/scala/org/apache/spark/sql/SQLImplicits.scala]: Current bug is that we are reading extra rows than what was originally cached. Could it be that this is caused because we rely on L278 to return 0?

File: sql/core/src/main/scala/org/apache/spark/sql/SQLImplicits.scala (hunk 8 above)
Comment (feynmanliang) [inline sql/core/src/main/scala/org/apache/spark/sql/SQLImplicits.scala]: :+1: , this would require explicitly writing a "-1" to indicate end of valid data (and a corresponding change in `totalRecordSize`)

File: sql/core/src/main/scala/org/apache/spark/sql/SQLImplicits.scala (hunk 7 above)
Comment (feynmanliang) [inline sql/core/src/main/scala/org/apache/spark/sql/SQLImplicits.scala]: I don't think I completely understand; `writeToMemory` delegates to `Platform.copyMemory` so are you referring to something else?

Comment (JoshRosen) [inline sql/core/src/main/scala/org/apache/spark/sql/SQLImplicits.scala]: I think that `convertToUnsafe.apply()` always returns the same object.

File: sql/core/src/main/scala/org/apache/spark/sql/SQLImplicits.scala [hunk 16]
```diff
@@ -110,4 +121,181 @@ private[sql] abstract class SQLImplicits {
...
+                  Platform.BYTE_ARRAY_OFFSET,
+                  rawBlock.size() - padding)
+
+                // Decompress into MemoryBlock backed by on-heap byte array
+                val compressedBais = new ByteArrayInputStream(compressedBlockArray)
+                val uncompressedBlockArray = new Array[Byte](_blockSize)
+                val cdis = new DataInputStream(codec.compressedInputStream(compressedBais))
+                // compression codec has its own block size and cis.read only reads one compression
+                // codec block, so we need to keep reading until cis is exhausted
+                // TODO: Use DataInputStream.readFully()
```
Comment (JoshRosen) [inline sql/core/src/main/scala/org/apache/spark/sql/SQLImplicits.scala]: Can use Guava's `ByteStreams.readFully()` instead.

**Comment by JoshRosen** [comment]
Just refactored this in a way that avoids unnecessary row format conversion; I think this makes the code easier to understand / reason about.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/TungstenCache.scala [hunk 17]
```diff
@@ -0,0 +1,179 @@
...
+    val childRDD: RDD[UnsafeRow] = {
+      if (child.outputsUnsafeRows) {
+        child.execute().asInstanceOf[RDD[UnsafeRow]]
+      } else {
+        ConvertToUnsafe(child).execute().asInstanceOf[RDD[UnsafeRow]]
+      }
+    }
+    val cachedRDD: RDD[MemoryBlock] = childRDD.mapPartitions { rowIterator =>
+      val bufferedRowIterator = rowIterator.buffered
+      val taskMemoryManager = TaskContext.get().taskMemoryManager()
```
Comment (JoshRosen) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/TungstenCache.scala]: I'm wondering what we actually gain by using the TaskMemoryManager here. The TMM would allow us to cache off-heap but it looks like we don't / won't currently do that if we use compression. For the initial version of this task, I'd like to focus on making the basic caching work really well and defer the question of how to integrate offheap caching with block manager to a followup PR.

Comment (JoshRosen) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/TungstenCache.scala]: Actually, I take that back: I see that we do actually end up allocating off-heap for the compressed data as well. We still have complexities in how to ensure cleanup when things are evicted, but that's out of scope for perf. benchmarking studies for now.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/TungstenCache.scala [hunk 18]
```diff
@@ -0,0 +1,179 @@
...
+      new Iterator[MemoryBlock] {
+        // NOTE: This assumes that size of every row < blockSize
+        def next(): MemoryBlock = {
+          // Packs rows into a `blockSize` bytes contiguous block of memory, starting a new block
+          // whenever the current fills up
+          // Each row is laid out in memory as [rowSize (4)|row (rowSize)]
+          val block = taskMemoryManager.allocateUnchecked(blockSize)
+
+          var currOffset = 0
+          val seenRows = ArrayBuffer[UnsafeRow]()
```
Comment (JoshRosen) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/TungstenCache.scala]: Ah, I see that this is just leftover debug code. I'll remove it.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/TungstenCache.scala [hunk 19]
```diff
@@ -0,0 +1,179 @@
...
+              codec.compressedInputStream(new ByteArrayInputStream(compressedBlockArray))
+            val decompressedBlock = new Array[Byte](blockSize)
+            ByteStreams.readFully(decompressionStream, decompressedBlock)
+            decompressionStream.close()
+            MemoryBlock.fromByteArray(decompressedBlock)
+          case None => rawBlock
+        }
+
+        val rows = new ArrayBuffer[UnsafeRow]()
+        var currOffset = 0
```
Comment (JoshRosen) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/TungstenCache.scala]: I'm going to make this into a long to avoid warnings about possible overflow when comparing to the block size (which is a long).

File: sql/core/src/main/scala/org/apache/spark/sql/execution/TungstenCache.scala [hunk 20]
```diff
@@ -0,0 +1,194 @@
...
+          private[this] def getNextRowSize(): Int = {
+            if (currOffset >= block.size()) {
+              -1
+            } else {
+              Platform.getInt(block.getBaseObject, block.getBaseOffset + currOffset)
+            }
+          }
+          override def hasNext: Boolean = _nextRowSize > 0
+          override def next(): UnsafeRow = {
+            // TODO: should probably have a null terminator rather than relying on zeroed out
```
Comment (feynmanliang) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/TungstenCache.scala]: This can be removed now

File: sql/core/src/main/scala/org/apache/spark/sql/execution/TungstenCache.scala [hunk 21]
```diff
@@ -0,0 +1,194 @@
...
+          var currOffset = 0
+          while (bufferedRowIterator.hasNext && currOffset < blockSize) {
+            val currRow = bufferedRowIterator.head
+            val recordSize = 4 + currRow.getSizeInBytes
+            if (currOffset + recordSize < blockSize) {
+              Platform.putInt(
+                block.getBaseObject, block.getBaseOffset + currOffset, currRow.getSizeInBytes)
+              currRow.writeToMemory(block.getBaseObject, block.getBaseOffset + currOffset + 4)
+              bufferedRowIterator.next()
+            }
```
Comment (feynmanliang) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/TungstenCache.scala]:
```scala
else {
  Platform.putInt(block.getBaseObject, block.getBaseOffset + currOffset, -1)
}
```

File: sql/core/src/main/scala/org/apache/spark/sql/execution/TungstenCache.scala [hunk 22]
```diff
@@ -0,0 +1,194 @@
...
+      val bufferedRowIterator = rowIterator.buffered
+      val taskMemoryManager = TaskContext.get().taskMemoryManager()
+      val compressionCodec: Option[CompressionCodec] =
+        compressionType.map { t => CompressionCodec.createCodec(SparkEnv.get.conf, t) }
+      new Iterator[MemoryBlock] {
+        // NOTE: This assumes that size of every row < blockSize
+        def next(): MemoryBlock = {
+          // Packs rows into a `blockSize` bytes contiguous block of memory, starting a new block
+          // whenever the current fills up.
+          // Each row is laid out in memory as [rowSize (int)|rowData (rowSize bytes)]
```
Comment (feynmanliang) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/TungstenCache.scala]: Layout is now [rowSize (int)|rowData (rowSize bytes)| "-1" (int)|xxxxx]

File: sql/core/src/main/scala/org/apache/spark/sql/execution/TungstenCache.scala [hunk 23]
```diff
@@ -0,0 +1,194 @@
...
+          val block = taskMemoryManager.allocateUnchecked(blockSize)
+
+          var currOffset = 0
+          while (bufferedRowIterator.hasNext && currOffset < blockSize) {
+            val currRow = bufferedRowIterator.head
+            val recordSize = 4 + currRow.getSizeInBytes
+            if (currOffset + recordSize < blockSize) {
```
Comment (feynmanliang) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/TungstenCache.scala]: `currOffset + recordSize < blockSize - 4` to account for the 4 bytes needed to insert terminator "-1"

File: sql/core/src/main/scala/org/apache/spark/sql/execution/TungstenCache.scala [hunk 24]
```diff
@@ -0,0 +1,194 @@
...
+      }
+    }
+    val cachedRDD: RDD[MemoryBlock] = childRDD.mapPartitions { rowIterator =>
+      // Note: this buffering is used to let us hold onto rows when rolling across block boundaries.
+      val bufferedRowIterator = rowIterator.buffered
+      val taskMemoryManager = TaskContext.get().taskMemoryManager()
+      val compressionCodec: Option[CompressionCodec] =
+        compressionType.map { t => CompressionCodec.createCodec(SparkEnv.get.conf, t) }
+      new Iterator[MemoryBlock] {
+        // NOTE: This assumes that size of every row < blockSize
```
Comment (feynmanliang) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/TungstenCache.scala]: `row < blockSize - 4`

File: sql/core/src/main/scala/org/apache/spark/sql/execution/TungstenCache.scala [hunk 25]
```diff
@@ -0,0 +1,194 @@
...
+              -1
+            } else {
+              Platform.getInt(block.getBaseObject, block.getBaseOffset + currOffset)
+            }
+          }
+          override def hasNext: Boolean = _nextRowSize > 0
+          override def next(): UnsafeRow = {
+            // TODO: should probably have a null terminator rather than relying on zeroed out
+            assert(_nextRowSize > 0)
+            currOffset += 4
```
Comment (feynmanliang) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/TungstenCache.scala]: Should document this is advancing offset over memory for "_nextRowSize"

File: sql/core/src/main/scala/org/apache/spark/sql/execution/TungstenCache.scala [hunk 26]
```diff
@@ -0,0 +1,194 @@
...
+      val baos = new ByteArrayOutputStream(memoryBlock.size().toInt)
+      val compressedBaos = compressionCodec.compressedOutputStream(baos)
+      compressedBaos.write(blockArray)
+      compressedBaos.flush()
+      compressedBaos.close()
+      baos.toByteArray
+    }
+
+    // Allocate a new block with compressed byte array padded to word boundary
+    val totalRecordSize = compressedBlockArray.length + 4
```
Comment (feynmanliang) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/TungstenCache.scala]: Should document the "+ 4" is to store amount of padding at head of block

**Comment by feynmanliang** [comment]
Looks much cleaner, thanks for refactoring!

File: sql/core/src/main/scala/org/apache/spark/sql/execution/TungstenCache.scala [hunk 27]
```diff
@@ -0,0 +1,194 @@
...
+          private[this] var currOffset: Long = 0
+          private[this] var _nextRowSize = getNextRowSize()
+          private[this] def getNextRowSize(): Int = {
+            if (currOffset >= block.size()) {
+              -1
+            } else {
+              Platform.getInt(block.getBaseObject, block.getBaseOffset + currOffset)
+            }
+          }
+          override def hasNext: Boolean = _nextRowSize > 0
```
Comment (feynmanliang) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/TungstenCache.scala]: After we add a "-1" terminator, `_nextRowSize` should in theory never `== 0`

**Comment by JoshRosen** [comment]
Just pushed a commit which begins to add an automated benchmarking harness. I used the UI REST APIs to automatically grab cached RDD storage statistics. Going to continue improving so that the columnar cache vs. Tungsten cache benchmark is fully-automated.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/TungstenCache.scala (hunk 27 above)
Comment (JoshRosen) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/TungstenCache.scala]: Yep, still planning to do this explicit terminator refactoring.

**Comment by feynmanliang** [comment]
I tried out the "-1" terminator code and am still getting the extra rows bug when `tungstenCache`ing `/mnt/performance-datasets/tpcds/sf1500/hasNullPartitionValue=false/shuffle=true/useDecimal=true/parquet/customer` on dogfood, so that wasn't it :/

File: sql/core/src/main/scala/org/apache/spark/sql/execution/TungstenCache.scala (hunk 21 above)
Comment (feynmanliang) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/TungstenCache.scala]: Actually, this will miss inserting a "-1" when `bufferedRowIterator` is exhausted

Comment (feynmanliang) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/TungstenCache.scala]:
```scala
          var currOffset = 0
          while (bufferedRowIterator.hasNext &&
                 currOffset + 4 + bufferedRowIterator.head.getSizeInBytes < blockSize - 4) {
            val currRow = bufferedRowIterator.head
              Platform.putInt(
                block.getBaseObject, block.getBaseOffset + currOffset, currRow.getSizeInBytes)
              currRow.writeToMemory(block.getBaseObject, block.getBaseOffset + currOffset + 4)
              bufferedRowIterator.next()
            currOffset += 4 + currRow.getSizeInBytes
          }
          // Write a -1 to indicate end of data
          Platform.putInt(block.getBaseObject, block.getBaseOffset + currOffset, -1)
```
should do it (passes unit tests, let's see what happens on dogfood)

Comment (JoshRosen) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/TungstenCache.scala]: We should add unit tests for caching empty partitions.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/TungstenCache.scala (hunk 24 above)
Comment (JoshRosen) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/TungstenCache.scala]: Actually, should this be -8 if we want to assume an end-of-block marker at the end of every page?

Comment (feynmanliang) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/TungstenCache.scala]: :+1: yep

File: sql/core/src/main/scala/org/apache/spark/sql/execution/TungstenCache.scala (hunk 21 above)
Comment (feynmanliang) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/TungstenCache.scala]: @JoshRosen @yhuai adding the -1 terminator and the above code fixes the extra row bug on dogfood

File: sql/core/src/main/scala/org/apache/spark/sql/execution/TungstenCache.scala [hunk 28]
```diff
@@ -0,0 +1,195 @@
...
+  }
+
+  private def decompressBlock(
+      compressedMemoryBlock: MemoryBlock,
+      decompressedSize: Int,
+      compressionCodec: CompressionCodec): MemoryBlock = {
+    // Copy compressed block (excluding padding) to on-heap byte array
+    val padding =
+      Platform.getInt(compressedMemoryBlock.getBaseObject, compressedMemoryBlock.getBaseOffset)
+    val compressedBlockArray = new Array[Byte](compressedMemoryBlock.size().toInt - padding)
```
Comment (feynmanliang) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/TungstenCache.scala]: I think @JoshRosen mentioned earlier that we don't need to reinitialize here and can reuse this `Array` across multiple blocks

File: sql/core/src/main/scala/org/apache/spark/sql/execution/TungstenCache.scala [hunk 29]
```diff
@@ -0,0 +1,195 @@
...
+      }
+    }
+  }
+
+  private def compressBlock(
+      memoryBlock: MemoryBlock,
+      compressionCodec: CompressionCodec,
+      taskMemoryManager: TaskMemoryManager): MemoryBlock = {
+    // Compress the block using an on-heap byte array
+    val compressedBlockArray: Array[Byte] = {
```
Comment (feynmanliang) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/TungstenCache.scala]: Can we also reuse this `Array` across `MemoryBlock`s since L162 copies it off-heap?
